### ISSUE TITLE
Change Wording of HaveCount to Highlight Mismatch of Expectation

### ISSUE THREAD

**Issue by Ben-G** [Quick/Nimble#308]
Hey folks!

First of all, thanks a lot for building Nimble! It's greatly improved the test quality on our project.

I was wondering if you would be interested in changing the failure message of the `haveCount` matcher.
Right now it prints the entire representation of the collection type along with the expected count.

This works well for the test cases that are part of Nimble, but in reality my collections tend to contain big data structures so it's always difficult to see the expected count and the actual count when a test fails (I usually need to open the issue navigator).

I was considering switching from:
```
expected to have [1,2,3] with count 1, got 3"
```

to simply using the type of the instance:

```
expected to have Array<Int> with count 1, got 3"
```

I would understand if you don't want this change. Just wanted to bring up the idea. Alternatively we could also append the detailed string representation of the data structure to the end of the failure message? That way we would retain the amount of detail we have now but we would be able to understand the failure without opening the issue navigator.

In case you like any of the approaches I'm happy to implement the change.

**Comment by ashfurrow** [comment]
Thanks for the question – could you clarify the problem with the issue navigator? Right now, this wording is consistent with other matchers (ex: "expected true to be false", "expected [] to not be empty", etc). Do you reckon there's  a way we can improve the message and the experience of quickly understanding the test failure without breaking the existing behaviour?

**Comment by Ben-G** [comment]
Thanks for the quick response. I should have added images to my issue. Hope these images help explain the issue:

**Current Implementation:**
<img width="1120" alt="screen shot 2016-06-22 at 8 27 06 am" src="https://cloud.githubusercontent.com/assets/1423068/16272835/2e0eea82-3854-11e6-89bd-fd53083e81ea.png">

**Potential Change:**
<img width="1103" alt="screen shot 2016-06-22 at 8 26 19 am" src="https://cloud.githubusercontent.com/assets/1423068/16272832/2c3292c2-3854-11e6-835f-c12bc7a57c16.png">

In the current implementation I find myself needing to open the issue navigator to find the information about actual count vs. expected count, because the description of the collection content takes up so much space.

Due to the way Xcode's issue navigator works this ends up taking a fair bit of time when navigating a very large test suite.

For expectations that use the `haveCount` matcher I'm usually only concerned about the element count and not the content of the array; so to me personally it would be more helpful to see the count values first.

That said, I see that this causes an inconsistency & the solution of using the `dynamicType` of the collection can get a little ugly when using Objective-C types, as it exposes types names of internal types (concrete subclasses of `NSArray`).

Just wondered if other people also see this as an issue.

Thanks a lot!

**Comment by ashfurrow** [comment]
That makes sense, thanks for the clarification. I'm not sure how to proceed, I think it's worth getting input from @Quick/contributors.

**Comment by modocache** [comment]
This sounds like a nice change to me! Thanks for the suggestion, @Ben-G. Would you be interested in sending a pull request to implement this?

It might be a good idea to add a source code comment to the implementation that links back to this issue -- I think this is an improvement, but some people might prefer more information, even if it is too long for Xcode to display.

**Comment by ashfurrow** [comment]
[![](https://imgs.xkcd.com/comics/workflow.png)](https://xkcd.com/1172/)

**Comment by Ben-G** [comment]
@modocache @ashfurrow Thanks for the quick response! I'll open a PR and add a link to this issue for future reference.

**Comment by Ben-G** [comment]
Implemented in #310 but never closed!
